### PR TITLE
Document what we mean by backward compatibility

### PR DESCRIPTION
**Pull request by gilles-peskine-arm**
Give some concrete examples of what is not covered by backward compatibility.

For example, according to this description, https://github.com/ARMmbed/mbedtls/issues/4167 is explicitly not a compatibility break (“Changing which error code is returned in a case where multiple error conditions apply”). Adding a new cryptographic algorithm to the default certificate acceptance profile isn't (“Changing the behavior of a function from failing to succeeding, when the change is a reasonable extension of the current behavior”). But removing an algorithm from the default profile would be (no success can change to an error unless it was undocumented behavior or security requires the change).

### CODE REVIEW

File: BRANCHES.md
Comment (gilles-peskine-arm) [inline BRANCHES.md:35]: What about changing items' values?

Comment (paul-elliott-arm) [reply BRANCHES.md:35]: Do we have any enums that are functionally dependant on the actual values of the enum? Presumably such items would have an explanatory comment next to them otherwise it becomes a maintenance nightmare. Perhaps we should caveat that adding items is only ok if changing the values of an enum has no effect.

Comment (gilles-peskine-arm) [reply BRANCHES.md:35]: Inside the library, I don't think there's any behavior that depends on the numerical values. But it could matter for people who write FFI and replicate enums as constant in the high-level language.

I think the rules should be the same for enum values and for numerical `#define`s. Many enum-like values are defines because we want to control the size of the storage or because the type is open-ended in the module that defines it.

Comment (paul-elliott-arm) [reply BRANCHES.md:35]: Definitely agree on the enums where they are used as #defines, I was pondering the case where adding an enum item to the middle of an enum causes renumbering of items after it, and some code is dependant on items having given values. That would not, admittedly be great code though.

Comment (daverodgman) [reply BRANCHES.md:35]: I think the case Paul mentions is an ABI break, obviously, but not an API break, on the basis that the API is defined using the enum names, not their values. So I don't think it would be completely wrong to add a bullet for "Changing existing enum values" - but that's not a common thing for us to do (as per the note preceding the list) so I would suggest we don't mention it.

File: BRANCHES.md
Comment (daverodgman) [inline BRANCHES.md:31]: Possibly marginally clearer to say "... not considered API compatibility breaks" (to distinguish from ABI concerns).

**Review by daverodgman — APPROVED** [review]

File: BRANCHES.md
Comment (yanesca) [inline BRANCHES.md:31]: Nitpick: Line wider than 80 columns.

File: BRANCHES.md
Comment (yanesca) [inline BRANCHES.md:38]: Nitpick: Lines wider than 80 columns.

File: BRANCHES.md
Comment (yanesca) [inline BRANCHES.md:40]: With the new section added it takes a bit of thinking to figure out to what rule are these exceptions meant. Maybe elaborating a bit would make it easier to read:
> There are rare exceptions when we break API compatibility:  code that was relying on something that became

File: BRANCHES.md
Comment (yanesca) [inline BRANCHES.md:28]: Pre-existing: maybe it would be helpful to separate API and the ABI compatibility discussions to their respective subsections.

**Review by mpg — COMMENTED** [review]
This looks pretty good to me, except for a few points already mentioned by Dave and Janos, which I've marked with :+1:

**Comment by mpg** [comment]
Note: a few other recurring questions / grey areas when it comes to compatibility is what exactly is covered by the promise:
- are the example programs covered? the test framework?
- what are our promises regarding versions of the tools needed to build the library? to build the tests? to configure the library (`scripts/config.py`, PSA Crypto drivers)? to hack on it (update generated files)?

I'm not sure if it's better to include that in this PR or to treat as separate issues. I definitely understand if you're rather avoid scope creep and keep this PR nice and small. I just wanted to mention those so that we can decide to handle them now or create issues to track them for later. (IMO those issues would fit "tech debt month" as it's a long-standing pain that we and our users don't know where we stand.)

(If we decide to go there, I'm inclined to restrict the our pretty strong guarantees to the library only, and promise much less regarding the rest, possibly nothing regarding programs initially - and perhaps introduce a list of "stable" programs later.)

**Comment by gilles-peskine-arm** [comment]
Ideally I'd like to define guarantees for end-user programs (but not test programs and sample programs) — which would entail first defining which programs are end-user programs. And I'd like to define guarantees regarding the build environment (tool requirements, script paths). But that's out of scope of this pull request. I made this PR to qualify our explicit API stability promise. For the rest, we currently don't make any explicit promises.

**Review by yanesca — APPROVED** [review]
LGTM.

**Review by mpg — APPROVED** [review]
Looks good to me as well.

File: BRANCHES.md
Comment (ronald-cron-arm) [inline BRANCHES.md:31]: Pre-existing but line 19 shouldn't it be: "we maintain
API compatibility in the `master` branch between MINOR version changes." ?

File: BRANCHES.md
Comment (ronald-cron-arm) [inline BRANCHES.md:38]:
```suggestion
* Returning an error code that was not previously documented for a function
```

Comment (gilles-peskine-arm) [reply BRANCHES.md]: I amended the last commit to restore the accidentally deleted `*`.

File: BRANCHES.md
Comment (gilles-peskine-arm) [reply BRANCHES.md:31]: Looks correct to me: between two changes of the major version (e.g. between the change from 1.3 to 2.0 and the change from 2.28 to 3.0), the API remains backward-compatible.

We maintain API compatibility *on* minor version changes, not just *between* minor version changes.

Comment (mpg) [reply BRANCHES.md:31]: I feel like "between" here can be interpreted in two ways:

1. "During the interval of time starting at 2.0.0 and ending before 3.0.0, we maintain..." -> _between_ those two points in time, we maintain... -> between major version changes, we maintain...
2. "Version 2.27 is backward compatible with 2.26" -> there's backwards compat _between_ those two versions -> between minor versions.

I think the fact that we're saying "between version changes" rather than "between versions" hints towards the first interpretation.

In general, I think it's important to reduce ambiguity, but in that case there's already a link to a detailed explanation, and we spell things in full in the first paragraph of the next section, so I'm not sure it's worth spending too much time on this.

Comment (ronald-cron-arm) [reply BRANCHES.md:31]: ok got it.

**Review by ronald-cron-arm — DISMISSED** [review]
I think this is a good improvement of the backward compatibility documentation thus approving.

File: BRANCHES.md
Comment (daverodgman) [reply BRANCHES.md:31]: I think this is misleading: I interpret saying "we maintain compatibility between major versions" as meaning, e.g., 2.x is API-compatible with 3.x. Maybe the following is clearer?

... we maintain API compatibility in the `master` branch across minor version changes, but not when the major version changes.

I'll approve as-is but would be in favour of addressing this.

**Review by daverodgman — APPROVED** [review]

**Comment by daverodgman** [comment]
Approved, but I would be in favour of addressing the point highlighted by Ronald on line 19 before merging. I'll leave it to another gatekeeper / Gilles to make that call.

File: BRANCHES.md
Comment (gilles-peskine-arm) [reply BRANCHES.md:32]: I've reworded the sentence to avoid the word “between” and added a formulation in terms of version numbers.

**Review by daverodgman — APPROVED** [review]
LGTM

**Review by ronald-cron-arm — APPROVED** [review]
Thanks for the last change on line 19. I find it much clearer now.

**Review by mpg — APPROVED** [review]
Still looks good to me!

**Pull request merged**
